src/backend/distributed/commands/schema_based_sharding.c

```diff
@@ -227,18 +643,212 @@ citus_internal_unregister_tenant_schema_globally(PG_FUNCTION_ARGS)
 	if (HeapTupleIsValid(namespaceTuple))
 	{
 		ReleaseSysCache(namespaceTuple);
-
 		ereport(ERROR, (errmsg("schema is expected to be already dropped "
 							   "because this function is only expected to "
 							   "be called from Citus drop hook")));
 	}
-
 	uint32 tenantSchemaColocationId = SchemaIdGetTenantColocationId(schemaId);
 
 	DeleteTenantSchemaLocally(schemaId);
 	SendCommandToWorkersWithMetadata(TenantSchemaDeleteCommand(schemaNameStr));
 
 	DeleteColocationGroup(tenantSchemaColocationId);
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * citus_schema_distribute gets a regular schema name, then converts it to a tenant
+ * schema.
+ */
+Datum
+citus_schema_distribute(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+	EnsureCoordinator();
+
+	Oid schemaId = PG_GETARG_OID(0);
+	EnsureSchemaOwner(schemaId);
+
+	/* Prevent concurrent schema deletion and table creation under the schema */
+	LockDatabaseObject(NamespaceRelationId, schemaId, 0, AccessExclusiveLock);
+
+	/*
+	 * It is possible that by the time we acquire the lock on schema,
+	 * concurrent DDL has removed it. We can test this by checking the
+	 * existence of schema.
+	 */
+	if (!SearchSysCacheExists1(NAMESPACEOID, ObjectIdGetDatum(schemaId)))
+	{
+		ereport(ERROR, (errcode(ERRCODE_UNDEFINED_SCHEMA),
+						errmsg("concurrent operation removed the schema")));
+	}
```

Comment:
We should check like below. Reasoning:

1. Malicious user cannot take the lock to block operations since we will throw error due to schema owner check and all the locks will be released,
2. Schema existence check should be done after taking the lock. Since schema can be seen as it exists due to cache, even if a concurrent operation removed it. But `LockDatabaseObject` calls `AcceptInvalidationMessages` which makes sure our cache is up to date after it azqures the lock for the schema.

```c
        /* Prevent concurrent table creation under the schema */
	LockDatabaseObject(NamespaceRelationId, schemaId, 0, AccessExclusiveLock);

	/* 
	 * We should ensure the existence of the schema after taking the lock since
	 * the schema could have been dropped before we acquired the lock.
	 */
	EnsureSchemaExist(schemaId);
	EnsureSchemaOwner(schemaId);
```